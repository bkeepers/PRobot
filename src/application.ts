import { WebhookEvent } from '@octokit/webhooks'
import deprecated from 'deprecated-decorator'
import express from 'express'
import { EventEmitter } from 'promise-events'
import { ApplicationFunction } from '.'
import { Context } from './context'
import { GitHubAPI } from './github'
import { GitHubApp } from './github-app'
import { logger } from './logger'
import { LoggerWithTarget, wrapLogger } from './wrap-logger'

/**
 * The `app` parameter available to apps
 *
 * @property {logger} log - A logger
 */
export class Application {
  public events: EventEmitter
  public router: express.Router
  public catchErrors: boolean
  public log: LoggerWithTarget

  private github: GitHubApp

  constructor (options?: Options) {
    const opts = options || {} as any
    this.events = new EventEmitter()
    this.log = wrapLogger(logger, logger)
    this.github = opts.github
    this.catchErrors = opts.catchErrors || false
    this.router = opts.router || express.Router() // you can do this?
  }

  /**
   * Loads a Probot plugin
   * @param plugin - Probot plugin to load
   */
  public load (app: ApplicationFunction | ApplicationFunction[]): Application {
    if (Array.isArray(app)) {
      app.forEach(a => this.load(a))
    } else {
      app(this)
    }

    return this
  }

  public async receive (event: WebhookEvent) {
    if ((event as any).event) {
      // tslint:disable-next-line:no-console
      console.warn(new Error('Propery `event` is deprecated, use `name`'))
      event = { name: (event as any).event, ...event }
    }

    return Promise.all([
      this.events.emit('*', event),
      this.events.emit(event.name, event),
      this.events.emit(`${ event.name }.${ event.payload.action }`, event)
    ])
  }

  /**
   * Get an {@link http://expressjs.com|express} router that can be used to
   * expose HTTP endpoints
   *
   * ```
   * module.exports = app => {
   *   // Get an express router to expose new HTTP endpoints
   *   const route = app.route('/my-app');
   *
   *   // Use any middleware
   *   route.use(require('express').static(__dirname + '/public'));
   *
   *   // Add a new route
   *   route.get('/hello-world', (req, res) => {
   *     res.end('Hello World');
   *   });
   * };
   * ```
   *
   * @param path - the prefix for the routes
   * @returns an [express.Router](http://expressjs.com/en/4x/api.html#router)
   */
  public route (path?: string): express.Router {
    if (path) {
      const router = express.Router()
      this.router.use(path, router)
      return router
    } else {
      return this.router
    }
  }

  /**
   * Listen for [GitHub webhooks](https://developer.github.com/webhooks/),
   * which are fired for almost every significant action that users take on
   * GitHub.
   *
   * @param event - the name of the [GitHub webhook
   * event](https://developer.github.com/webhooks/#events). Most events also
   * include an "action". For example, the * [`issues`](
   * https://developer.github.com/v3/activity/events/types/#issuesevent)
   * event has actions of `assigned`, `unassigned`, `labeled`, `unlabeled`,
   * `opened`, `edited`, `milestoned`, `demilestoned`, `closed`, and `reopened`.
   * Often, your bot will only care about one type of action, so you can append
   * it to the event name with a `.`, like `issues.closed`.
   *
   * ```js
   * app.on('push', context => {
   *   // Code was just pushed.
   * });
   *
   * app.on('issues.opened', context => {
   *   // An issue was just opened.
   * });
   * ```
   *
   * @param callback - a function to call when the
   * webhook is received.
   */
  public on (eventName: string | string[], callback: (context: Context) => Promise<void>) {
    if (typeof eventName === 'string') {
      return this.events.on(eventName, async (event: WebhookEvent) => {
        try {
          await callback(await this.github.createContext(event))
        } catch (err) {
          this.log.error({ err, event, id: event.id })
          if (!this.catchErrors) {
            throw err
          }
        }
      })
    } else {
      eventName.forEach(e => this.on(e, callback))
    }
  }

  @deprecated('github.jwt')
  public app (): string {
    return this.github.jwt()
  }

  @deprecated('github.auth')
  public auth (id?: number, log = this.log): Promise<GitHubAPI> {
    return this.github.auth(id, log)
  }
}

export interface Options {
  github: GitHubApp
  router?: express.Router
  catchErrors?: boolean
}
