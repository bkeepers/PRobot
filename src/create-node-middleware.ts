import { RequestListener, IncomingMessage, ServerResponse } from "http";
import { NextFunction } from "express";

import { ApplicationFunction } from "./types";
import { MiddlewareOptions } from "./types";

export function createNodeMiddleware(
  appFn: ApplicationFunction,
  { probot }: MiddlewareOptions
): RequestListener {
  return (
    request: IncomingMessage,
    response: ServerResponse,
    next?: NextFunction
  ) => {
    probot.load(appFn);
    probot.webhooks.middleware(request, response, next);
  };
}
