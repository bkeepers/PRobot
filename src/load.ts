import { Deprecation } from "deprecation";
import { Router } from "express";

import { Application } from "./application";
import { ApplicationFunction, ApplicationFunctionOptions } from "./types";
import { getRouter } from "./get-router";

type DeprecatedKey = "auth" | "load" | "log" | "on" | "receive" | "router";

const DEPRECATED_APP_KEYS: DeprecatedKey[] = [
  "auth",
  "load",
  "log",
  "on",
  "receive",
  "router",
];
let didDeprecate = false;

/**
 * Loads an ApplicationFunction into the current Application
 * @param appFn - Probot application function to load
 */
export function load(
  app: Application,
  router: Router | null,
  appFn: ApplicationFunction | ApplicationFunction[]
) {
  const deprecatedApp = DEPRECATED_APP_KEYS.reduce(
    (api: Record<string, unknown>, key: DeprecatedKey) => {
      Object.defineProperty(api, key, {
        get() {
          if (didDeprecate) return app[key];

          app.log.warn(
            new Deprecation(
              '[probot] "(app) => {}" is deprecated. Use "({ app }) => {}" instead'
            )
          );
          didDeprecate = true;

          return app[key];
        },
      });

      return api;
    },
    {}
  );

  if (Array.isArray(appFn)) {
    appFn.forEach((fn) => load(app, router, fn));
  } else {
    appFn(
      (Object.assign(deprecatedApp, {
        app,
        getRouter: getRouter.bind(null, router || app.router),
      }) as unknown) as ApplicationFunctionOptions
    );
  }

  return app;
}
