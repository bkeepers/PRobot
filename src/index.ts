// tslint:disable-next-line: no-var-requires
require("dotenv").config();

import { Server } from "http";

import express from "express";
import Redis from "ioredis";
import LRUCache from "lru-cache";
import { Deprecation } from "deprecation";
import pinoHttp from "pino-http";
import { getPrivateKey } from "@probot/get-private-key";

import type { WebhookEvent, Webhooks } from "@octokit/webhooks";
import type { Logger } from "pino";

import { Application } from "./application";
import { setupAppFactory } from "./apps/setup";
import { Context, WebhookPayloadWithRepository } from "./context";
import { ProbotOctokit } from "./octokit/probot-octokit";
import { getLog } from "./helpers/get-log";
import { resolveAppFunction } from "./helpers/resolve-app-function";
import { createServer } from "./server/create-server";
import { createWebhookProxy } from "./helpers/webhook-proxy";
import { getErrorHandler } from "./helpers/get-error-handler";
import {
  ApplicationFunction,
  DeprecatedLogger,
  ProbotWebhooks,
  State,
} from "./types";
import { getProbotOctokitWithDefaults } from "./octokit/get-probot-octokit-with-defaults";
import { aliasLog } from "./helpers/alias-log";
import { logWarningsForObsoleteEnvironmentVariables } from "./helpers/log-warnings-for-obsolete-environment-variables";
import { getWebhooks } from "./octokit/get-webhooks";
import { webhookEventCheck } from "./helpers/webhook-event-check";
import { auth } from "./auth";
import { load } from "./load";
import { getRouter } from "./get-router";

logWarningsForObsoleteEnvironmentVariables();

export interface Options {
  // same options as Application class
  privateKey?: string;
  githubToken?: string;
  id?: number;
  Octokit?: typeof ProbotOctokit;
  log?: Logger;
  redisConfig?: Redis.RedisOptions;
  secret?: string;
  webhookPath?: string;

  // Probot class-specific options
  /**
   * @deprecated `cert` options is deprecated. Use `privateKey` instead
   */
  cert?: string;
  port?: number;
  host?: string;
  webhookProxy?: string;
  /**
   * @deprecated set `Octokit` to `ProbotOctokit.defaults({ throttle })` instead
   */
  throttleOptions?: any;
}

// tslint:disable:no-var-requires
const defaultAppFns: ApplicationFunction[] = [require("./apps/default")];
// tslint:enable:no-var-requires

export class Probot {
  public static async run(appFn: ApplicationFunction | string[]) {
    const pkgConf = require("pkg-conf");
    const program = require("commander");

    const readOptions = (): Options => {
      if (Array.isArray(appFn)) {
        program
          .usage("[options] <apps...>")
          .option(
            "-p, --port <n>",
            "Port to start the server on",
            process.env.PORT || 3000
          )
          .option(
            "-H --host <host>",
            "Host to start the server on",
            process.env.HOST
          )
          .option(
            "-W, --webhook-proxy <url>",
            "URL of the webhook proxy service.`",
            process.env.WEBHOOK_PROXY_URL
          )
          .option(
            "-w, --webhook-path <path>",
            "URL path which receives webhooks. Ex: `/webhook`",
            process.env.WEBHOOK_PATH
          )
          .option("-a, --app <id>", "ID of the GitHub App", process.env.APP_ID)
          .option(
            "-s, --secret <secret>",
            "Webhook secret of the GitHub App",
            process.env.WEBHOOK_SECRET
          )
          .option(
            "-P, --private-key <file>",
            "Path to certificate of the GitHub App",
            process.env.PRIVATE_KEY_PATH
          )
          .parse(appFn);

        return {
          privateKey:
            getPrivateKey({ filepath: program.privateKey }) || undefined,
          id: program.app,
          port: program.port,
          host: program.host,
          secret: program.secret,
          webhookPath: program.webhookPath,
          webhookProxy: program.webhookProxy,
        };
      }
      const privateKey = getPrivateKey();
      return {
        privateKey: (privateKey && privateKey.toString()) || undefined,
        id: Number(process.env.APP_ID),
        port: Number(process.env.PORT) || 3000,
        host: process.env.HOST,
        secret: process.env.WEBHOOK_SECRET,
        webhookPath: process.env.WEBHOOK_PATH,
        webhookProxy: process.env.WEBHOOK_PROXY_URL,
      };
    };

    const options = readOptions();
    const probot = new Probot(options);

    if (!options.id || !options.privateKey) {
      if (process.env.NODE_ENV === "production") {
        if (!options.id) {
          throw new Error(
            "Application ID is missing, and is required to run in production mode. " +
              "To resolve, ensure the APP_ID environment variable is set."
          );
        } else if (!options.privateKey) {
          throw new Error(
            "Certificate is missing, and is required to run in production mode. " +
              "To resolve, ensure either the PRIVATE_KEY or PRIVATE_KEY_PATH environment variable is set and contains a valid certificate"
          );
        }
      }
      probot.load(setupAppFactory(probot.options.host, probot.options.port));
    } else if (Array.isArray(appFn)) {
      const pkg = await pkgConf("probot");
      probot.setup(program.args.concat(pkg.apps || []));
    } else {
      probot.load(appFn);
    }
    probot.start();

    return probot;
  }

  public server: express.Application;
  public webhooks: ProbotWebhooks;
  public log: DeprecatedLogger;
  public version: String;
  public on: ProbotWebhooks["on"];
  public auth: (
    installationId?: number,
    log?: Logger
  ) => Promise<InstanceType<typeof ProbotOctokit>>;

  // These need to be public for the tests to work.
  public options: Options;
  public throttleOptions: any;

  private httpServer?: Server;
  private state: State;

  /**
   * @deprecated this.internalRouter can be removed once we remove the Application class
   */
  private internalRouter: express.Router;

  /**
   * @deprecated use probot.log instead
   */
  public get logger() {
    this.log.warn(
      new Deprecation(
        `[probot] "probot.logger" is deprecated. Use "probot.log" instead`
      )
    );
    return this.log;
  }

  /**
   * @deprecated "app.router" is deprecated, use "getRouter()" from the app function instead: "({ app, getRouter }) => { ... }"
   */
  public get router() {
    this.log.warn(
      new Deprecation(
        `[probot] "app.router" is deprecated, use "getRouter()" from the app function instead: "({ app, getRouter }) => { ... }"`
      )
    );

    return this.internalRouter;
  }

  constructor(options: Options) {
    if (process.env.GHE_HOST && /^https?:\/\//.test(process.env.GHE_HOST)) {
      throw new Error(
        "Your `GHE_HOST` environment variable should not begin with https:// or http://"
      );
    }

    //
    // Probot class-specific options (Express server & Webhooks)
    //
    options.webhookPath = options.webhookPath || "/";
    options.secret = options.secret || "development";

    this.log = aliasLog(options.log || getLog());

    if (options.cert) {
      this.log.warn(
        new Deprecation(
          `[probot] "cert" option is deprecated. Use "privateKey" instead`
        )
      );
      options.privateKey = options.cert;
    }

    if (process.env.INSTALLATION_TOKEN_TTL) {
      this.log.warn(
        '[probot] "INSTALLATION_TOKEN_TTL" environment variable is no longer used. Tokens are renewed as needed at the time of the request now.'
      );
    }

    // TODO: support redis backend for access token cache if `options.redisConfig || process.env.REDIS_URL`
    const cache = new LRUCache<number, string>({
      // cache max. 15000 tokens, that will use less than 10mb memory
      max: 15000,
      // Cache for 1 minute less than GitHub expiry
      maxAge: 1000 * 60 * 59,
    });

    const Octokit = getProbotOctokitWithDefaults({
      githubToken: options.githubToken,
      Octokit: options.Octokit || ProbotOctokit,
      appId: options.id,
      privateKey: options.privateKey,
      cache,
      log: this.log,
      redisConfig: options.redisConfig,
      throttleOptions: options.throttleOptions,
    });
    const octokit = new Octokit();

    this.state = {
      cache,
      githubToken: options.githubToken,
      log: this.log,
      Octokit,
      octokit,
      webhooks: {
        path: options.webhookPath,
        secret: options.secret,
      },
      id: options.id,
      privateKey: options.privateKey,
    };

    this.auth = auth.bind(null, this.state);

    this.webhooks = getWebhooks(this.state);

    this.on = (eventNameOrNames, callback) => {
      // when an app subscribes to an event using `app.on(event, callback)`, Probot sends a request to `GET /app` and
      // verifies if the app is subscribed to the event and logs a warning if it is not.
      //
      // This feature will be moved out of Probot core as it has side effects and does not work in a stateless environment.
      webhookEventCheck(this.state, eventNameOrNames);

      if (eventNameOrNames === "*") {
        // @ts-ignore this workaround is only to surpress a warning. The `.on()` method will be deprecated soon anyway.
        return this.webhooks.onAny(callback);
      }

      return this.webhooks.on(eventNameOrNames, callback);
    };

    this.server = createServer({
      webhook: (this.webhooks as any).middleware,
      logger: this.log,
    });

    const { version } = require("../package.json");
    this.version = version;

    // TODO: remove once Application class was removed
    this.internalRouter = express.Router();

    // TODO: Refactor tests so we we can remove these
    this.options = options;
  }

  /**
   * @deprecated `probot.webhook` is deprecated. Use `probot.webhooks` instead
   */
  public get webhook(): Webhooks {
    this.log.warn(
      new Deprecation(
        `[probot] "probot.webhook" is deprecated. Use "probot.webhooks" instead`
      )
    );

    return this.webhooks;
  }

  public receive(event: WebhookEvent) {
    this.log.debug({ event }, "Webhook received");
    return this.webhooks.receive(event);
  }

  public load(appFn: string | ApplicationFunction | ApplicationFunction[]) {
    if (typeof appFn === "string") {
      appFn = resolveAppFunction(appFn) as ApplicationFunction;
    }

    const router = express.Router();

    // Connect the router from the app to the server
    this.server.use(router);

    // Initialize the ApplicationFunction
    load(this, router, appFn);

    return this;
  }

  public setup(appFns: Array<string | ApplicationFunction>) {
    // Log all unhandled rejections
    (process as NodeJS.EventEmitter).on(
      "unhandledRejection",
      getErrorHandler(this.log)
    );

    // Load the given appFns along with the default ones
    appFns.concat(defaultAppFns).forEach((appFn) => this.load(appFn));

    // Register error handler as the last middleware
    this.server.use(
      pinoHttp({
        logger: this.log,
      })
    );
  }

  public start() {
    this.log.info(
      `Running Probot v${this.version} (Node.js: ${process.version})`
    );
    const port = this.options.port || 3000;
    const { host } = this.options;
    const printableHost = host ?? "localhost";

    this.httpServer = this.server
      .listen(port, ...((host ? [host] : []) as any), () => {
        if (this.options.webhookProxy) {
          createWebhookProxy({
            logger: this.log,
            path: this.options.webhookPath,
            port: this.options.port,
            url: this.options.webhookProxy,
          });
        }
        this.log.info(`Listening on http://${printableHost}:${port}`);
      })
      .on("error", (error: NodeJS.ErrnoException) => {
        if (error.code === "EADDRINUSE") {
          this.log.error(
            `Port ${this.options.port} is already in use. You can define the PORT environment variable to use a different port.`
          );
        } else {
          this.log.error(error);
        }
        process.exit(1);
      });

    return this.httpServer;
  }

  public stop() {
    if (!this.httpServer) return;

    this.httpServer.close();
  }

  /**
   * Get an {@link http://expressjs.com|express} router that can be used to
   * expose HTTP endpoints
   *
   * ```
   * module.exports = ({ app, getRouter }) => {
   *   // Get an express router to expose new HTTP endpoints
   *   const router = getRouter('/my-app');
   *
   *   // Use any middleware
   *   router.use(require('express').static(__dirname + '/public'));
   *
   *   // Add a new route
   *   router.get('/hello-world', (req, res) => {
   *     res.end('Hello World');
   *   });
   * };
   * ```
   *
   * @param path - the prefix for the routes* @param path
   *
   * @deprecated "app.route()" is deprecated, use the "getRouter()" argument from the app function instead: "({ app, getRouter }) => { ... }"
   */
  route(path?: string) {
    this.log.warn(
      new Deprecation(
        `[probot] "app.route()" is deprecated, use the "getRouter()" argument from the app function instead: "({ app, getRouter }) => { ... }"`
      )
    );

    return getRouter(this.internalRouter, path);
  }
}

export const createProbot = (options: Options) => {
  options.log = options.log || getLog();
  options.log.warn(
    new Deprecation(
      `[probot] "createProbot(options)" is deprecated, use "new Probot(options)" instead`
    )
  );
  return new Probot(options);
};

export { Logger, Context, Application, ProbotOctokit };

/** NOTE: exported types might change at any point in time */
export { WebhookPayloadWithRepository };
