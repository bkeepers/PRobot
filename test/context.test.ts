import fs = require("fs");
import path = require("path");

import { WebhookEvent } from "@octokit/webhooks";
import nock from "nock";

import { Context } from "../src";
import { ProbotOctokit } from "../src/octokit/probot-octokit";

describe("Context", () => {
  let event: WebhookEvent;
  let context: Context;

  beforeEach(() => {
    event = {
      id: "123",
      name: "push",
      payload: {
        issue: { number: 4 },
        repository: {
          name: "probot",
          owner: { login: "bkeepers" },
        },
      },
    };

    context = new Context(event, {} as any, {} as any);
  });

  it("inherits the payload", () => {
    expect(context.payload).toBe(event.payload);
  });

  it("aliases the event name", () => {
    expect(context.name).toEqual("push");
    expect(context.event).toEqual("push");
  });

  describe("repo", () => {
    it("returns attributes from repository payload", () => {
      expect(context.repo()).toEqual({ owner: "bkeepers", repo: "probot" });
    });

    it("merges attributes", () => {
      expect(context.repo({ foo: 1, bar: 2 })).toEqual({
        bar: 2,
        foo: 1,
        owner: "bkeepers",
        repo: "probot",
      });
    });

    it("overrides repo attributes", () => {
      expect(context.repo({ owner: "muahaha" })).toEqual({
        owner: "muahaha",
        repo: "probot",
      });
    });

    // The `repository` object on the push event has a different format than the other events
    // https://developer.github.com/v3/activity/events/types/#pushevent
    it("properly handles the push event", () => {
      event.payload = require("./fixtures/webhook/push");

      context = new Context(event, {} as any, {} as any);
      expect(context.repo()).toEqual({ owner: "bkeepers-inc", repo: "test" });
    });

    it("return error for context.repo() when repository doesn't exist", () => {
      delete context.payload.repository;
      try {
        context.repo();
      } catch (e) {
        expect(e.message).toMatch("context.repo() is not supported");
      }
    });
  });

  describe("issue", () => {
    it("returns attributes from repository payload", () => {
      expect(context.issue()).toEqual({
        owner: "bkeepers",
        repo: "probot",
        issue_number: 4,
      });
    });

    it("merges attributes", () => {
      expect(context.issue({ foo: 1, bar: 2 })).toEqual({
        bar: 2,
        foo: 1,
        issue_number: 4,
        owner: "bkeepers",
        repo: "probot",
      });
    });

    it("overrides repo attributes", () => {
      expect(context.issue({ owner: "muahaha", issue_number: 5 })).toEqual({
        issue_number: 5,
        owner: "muahaha",
        repo: "probot",
      });
    });
  });

  describe("pullRequest", () => {
    it("returns attributes from repository payload", () => {
      expect(context.pullRequest()).toEqual({
        owner: "bkeepers",
        repo: "probot",
        pull_number: 4,
      });
    });

    it("merges attributes", () => {
      expect(context.pullRequest({ foo: 1, bar: 2 })).toEqual({
        bar: 2,
        foo: 1,
        owner: "bkeepers",
        pull_number: 4,
        repo: "probot",
      });
    });

    it("overrides repo attributes", () => {
      expect(context.pullRequest({ owner: "muahaha", pull_number: 5 })).toEqual(
        {
          owner: "muahaha",
          pull_number: 5,
          repo: "probot",
        }
      );
    });
  });

  describe("config", () => {
    let github: InstanceType<typeof ProbotOctokit>;

    function nockConfigResponseDataFile(fileName: string) {
      const configPath = path.join(__dirname, "fixtures", "config", fileName);
      const content = fs.readFileSync(configPath, { encoding: "utf8" });

      return {
        content: Buffer.from(content).toString("base64"),
      };
    }

    beforeEach(() => {
      github = new ProbotOctokit({
        retry: { enabled: false },
        throttle: { enabled: false },
      });
      context = new Context(event, github, {} as any);
    });

    it("gets a valid configuration", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, nockConfigResponseDataFile("basic.yml"));

      const config = await context.config("test-file.yml");
      expect(config).toEqual({
        bar: 7,
        baz: 11,
        foo: 5,
      });
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("returns null when the file and base repository are missing", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(404)
        .get("/repos/bkeepers/.github/contents/.github%2Ftest-file.yml")
        .reply(404);

      expect(await context.config("test-file.yml")).toBe(null);
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("returns the default config when the file and base repository are missing and default config is passed", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(404)
        .get("/repos/bkeepers/.github/contents/.github%2Ftest-file.yml")
        .reply(404);

      const defaultConfig = {
        bar: 7,
        baz: 11,
        foo: 5,
      };
      const contents = await context.config("test-file.yml", defaultConfig);
      expect(contents).toEqual(defaultConfig);
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("merges the default config", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, nockConfigResponseDataFile("basic.yml"));

      const config = await context.config("test-file.yml", { bar: 1, boa: 6 });
      expect(config).toEqual({
        bar: 7,
        baz: 11,
        boa: 6,
        foo: 5,
      });
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("merges a base config", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, {
          content: Buffer.from("boa: 6\nfoo: 0\n_extends: base").toString(
            "base64"
          ),
        })
        .get("/repos/bkeepers/base/contents/.github%2Ftest-file.yml")
        .reply(200, nockConfigResponseDataFile("basic.yml"));

      const config = await context.config("test-file.yml", { bar: 1, boa: 6 });

      expect(config).toEqual({
        bar: 7,
        baz: 11,
        boa: 6,
        foo: 0,
      });
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("merges the base and default config", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, {
          content: Buffer.from("boa: 6\nfoo: 0\n_extends: base").toString(
            "base64"
          ),
        })
        .get("/repos/bkeepers/base/contents/.github%2Ftest-file.yml")
        .reply(200, nockConfigResponseDataFile("basic.yml"));

      const config = await context.config("test-file.yml", {
        bar: 1,
        new: true,
      });
      expect(config).toEqual({
        bar: 7,
        baz: 11,
        boa: 6,
        foo: 0,
        new: true,
      });
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("merges a base config from another organization", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, {
          content: Buffer.from("boa: 6\nfoo: 0\n_extends: other/base").toString(
            "base64"
          ),
        })
        .get("/repos/other/base/contents/.github%2Ftest-file.yml")
        .reply(200, nockConfigResponseDataFile("basic.yml"));

      const config = await context.config("test-file.yml");

      expect(config).toEqual({
        bar: 7,
        baz: 11,
        boa: 6,
        foo: 0,
      });
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("merges a base config with a custom path", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, {
          content: Buffer.from(
            "boa: 6\nfoo: 0\n_extends: base:test.yml"
          ).toString("base64"),
        })
        .get("/repos/bkeepers/base/contents/test.yml")
        .reply(200, nockConfigResponseDataFile("basic.yml"));

      const config = await context.config("test-file.yml");
      expect(config).toEqual({
        bar: 7,
        baz: 11,
        boa: 6,
        foo: 0,
      });
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("ignores a missing base config", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, {
          content: Buffer.from("boa: 6\nfoo: 0\n_extends: base").toString(
            "base64"
          ),
        })
        .get("/repos/bkeepers/base/contents/.github%2Ftest-file.yml")
        .reply(404);

      const config = await context.config("test-file.yml");

      expect(config).toEqual({
        boa: 6,
        foo: 0,
      });
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("throws when the configuration file is malformed", async () => {
      expect.assertions(2);

      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, nockConfigResponseDataFile("malformed.yml"));

      try {
        await context.config("test-file.yml");
      } catch (error) {
        expect(error.message).toMatch(
          /^end of the stream or a document separator/
        );
        expect(mock.activeMocks()).toStrictEqual([]);
      }
    });

    it("throws when loading unsafe yaml", async () => {
      expect.assertions(2);

      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, nockConfigResponseDataFile("evil.yml"));

      try {
        await context.config("test-file.yml");
      } catch (error) {
        expect(error.message).toMatch(/unknown tag/);
        expect(mock.activeMocks()).toStrictEqual([]);
      }
    });

    it("throws on a non-string base", async () => {
      expect.assertions(2);

      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, {
          content: Buffer.from("boa: 6\nfoo: 0\n_extends: { nope }").toString(
            "base64"
          ),
        });

      try {
        await context.config("test-file.yml");
      } catch (error) {
        expect(error.message).toMatch(/invalid/i);
        expect(mock.activeMocks()).toStrictEqual([]);
      }
    });

    it("throws on an invalid base", async () => {
      expect.assertions(2);

      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, {
          content: Buffer.from('boa: 6\nfoo: 0\n_extends: "nope:"').toString(
            "base64"
          ),
        });

      try {
        await context.config("test-file.yml");
      } catch (error) {
        expect(error.message).toMatch(/nope:/);
        expect(mock.activeMocks()).toStrictEqual([]);
      }
    });

    it("returns an empty object when the file is empty", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, nockConfigResponseDataFile("empty.yml"));

      const contents = await context.config("test-file.yml");

      expect(contents).toEqual({});
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("overwrites default config settings", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, nockConfigResponseDataFile("basic.yml"));

      const config = await context.config("test-file.yml", { foo: 10 });

      expect(config).toEqual({
        bar: 7,
        baz: 11,
        foo: 5,
      });
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("uses default settings to fill in missing options", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, nockConfigResponseDataFile("basic.yml"));

      const config = await context.config("test-file.yml", { bar: 7 });

      expect(config).toEqual({
        bar: 7,
        baz: 11,
        foo: 5,
      });
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("uses the .github directory on a .github repo", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, {
          content: Buffer.from("foo: foo\n_extends: .github").toString(
            "base64"
          ),
        })
        .get("/repos/bkeepers/.github/contents/.github%2Ftest-file.yml")
        .reply(200, nockConfigResponseDataFile("basic.yml"));

      const config = await context.config("test-file.yml");

      expect(config).toEqual({
        bar: 7,
        baz: 11,
        foo: "foo",
      });
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("defaults to .github repo if no config found", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(404)
        .get("/repos/bkeepers/.github/contents/.github%2Ftest-file.yml")
        .reply(200, nockConfigResponseDataFile("basic.yml"));

      const config = await context.config("test-file.yml");

      expect(config).toEqual({
        bar: 7,
        baz: 11,
        foo: 5,
      });
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("deep merges the base config", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, {
          content: Buffer.from(
            "obj:\n  foo:\n  - name: master\n_extends: .github"
          ).toString("base64"),
        })
        .get("/repos/bkeepers/.github/contents/.github%2Ftest-file.yml")
        .reply(200, {
          content: Buffer.from("obj:\n  foo:\n  - name: develop").toString(
            "base64"
          ),
        });

      const config = await context.config("test-file.yml");

      expect(config).toEqual({
        obj: {
          foo: [{ name: "develop" }, { name: "master" }],
        },
      });
      expect(mock.activeMocks()).toStrictEqual([]);
    });

    it("accepts deepmerge options", async () => {
      const mock = nock("https://api.github.com")
        .get("/repos/bkeepers/probot/contents/.github%2Ftest-file.yml")
        .reply(200, {
          content: Buffer.from(
            "foo:\n  - name: master\n    shouldChange: changed\n_extends: .github"
          ).toString("base64"),
        })
        .get("/repos/bkeepers/.github/contents/.github%2Ftest-file.yml")
        .reply(200, {
          content: Buffer.from(
            "foo:\n  - name: develop\n  - name: master\n    shouldChange: should"
          ).toString("base64"),
        });

      const customMerge = jest.fn(
        (_target: any[], _source: any[], _options: any): any[] => []
      );
      await context.config("test-file.yml", {}, { arrayMerge: customMerge });
      expect(customMerge).toHaveBeenCalled();
      expect(mock.activeMocks()).toStrictEqual([]);
    });
  });

  describe("isBot", () => {
    test("returns true if sender is a bot", () => {
      event.payload.sender = { type: "Bot" };
      context = new Context(event, {} as any, {} as any);

      expect(context.isBot).toBe(true);
    });

    test("returns false if sender is not a bot", () => {
      event.payload.sender = { type: "User" };
      context = new Context(event, {} as any, {} as any);

      expect(context.isBot).toBe(false);
    });
  });
});
