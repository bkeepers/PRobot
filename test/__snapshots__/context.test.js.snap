// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context repo returns error for context.repo() 1`] = `"context.repo() is not supported for this webhook event."`;
